collector/mdadm.go

```diff
@@ -0,0 +1,263 @@
+	if len(matches) < 1+1 {
+		return 0, fmt.Errorf("too few matches found in buildline: %s", buildline)
+	} else {
+		if len(matches) > 1+1 {
+			return 0, fmt.Errorf("too many matches found in buildline: %s", buildline)
+		}
+	}
+
+	syncedSize, err := strconv.ParseInt(matches[1], 10, 64)
+
+	if err != nil {
+		return 0, fmt.Errorf("%s in buildline: %s", err, buildline)
+	} else {
+		return syncedSize, nil
+	}
+}
+
+// Parses an mdstat-file and returns a struct with the relevant infos.
+func parseMdstat(mdstatusfilepath string) ([]mdstatus, error) {
+	content, err := ioutil.ReadFile(mdstatusfilepath)
+	if err != nil {
+		return []mdstatus{}, fmt.Errorf("error parsing %s: %s", statusfile, err)
+	}
+
+	mdStatusFile := string(content)
+
+	lines := strings.Split(mdStatusFile, "\n")
+	var currentMD string
+
+	// Each md has at least the deviceline, statusline and one empty line afterwards
+	// so we will have probably something of the order len(lines)/3 devices
+	// so we use that for preallocation.
+	estimateMds := len(lines) / 3
+	mdstates := make([]mdstatus, 0, estimateMds)
+
+	for i, l := range lines {
+		if l == "" {
+			// Skip entirely empty lines.
+			continue
+		}
+
+		if l[0] == ' ' {
+			// Those lines are not the beginning of a md-section.
+			continue
+		} else {
+
+			if strings.HasPrefix(l, "Personalities") || strings.HasPrefix(l, "unused") {
+				// We aren't interested in lines with general info.
+				continue
+			}
+
+			mainLine := strings.Split(l, " ")
+			currentMD = mainLine[0]               // name of md-device
+			isActive := (mainLine[2] == "active") // activity status of said md-device
+
+			active, total, size, err := evalStatusline(lines[i+1]) // parse statusline, always present
+
+			if err != nil {
+				return mdstates, fmt.Errorf("error parsing %s: %s", statusfile, err)
+			}
+
+			// Now get the number of synced blocks.
+			var syncedBlocks int64
+
+			// Get the line number of the syncing-line.
+			var j int
+			if strings.Contains(lines[i+2], "bitmap") { // then skip the bitmap line
+				j = i + 3
+			} else {
+				j = i + 2
+			}
+
+			// If device is syncing at the moment, get the number of currently synced bytes,
+			// otherwise that number equals the size of the device.
+			if strings.Contains(lines[j], "recovery") || strings.Contains(lines[j], "resync") {
+				syncedBlocks, err = evalBuildline(lines[j])
+				if err != nil {
+					return mdstates, fmt.Errorf("error parsing %s: %s", statusfile, err)
+				}
+			} else {
+				syncedBlocks = size
+			}
+
+			mdstates = append(mdstates, mdstatus{currentMD, isActive, active, total, size, syncedBlocks})
+
+		}
+
+	}
+
+	return mdstates, nil
+}
+
+// Just returns the pointer to an empty struct as we only use throwaway-metrics.
+func NewMdadmCollector() (Collector, error) {
+	return &mdadmCollector{}, nil
+}
+
+var (
+	isActiveDesc = prometheus.NewDesc(
+			prometheus.BuildFQName("", "md", "is_active"),
```

Comment:
The first argument should always be `Namespace` here. It is defined here in the same package: https://github.com/prometheus/node_exporter/blob/master/collector/collector.go#L8